File: src/test/java/one/edee/babylon/export/MessageFileProcessorTest.java
package one.edee.babylon.export;

import one.edee.babylon.export.MessageFileProcessor.Pair;
import one.edee.babylon.export.MessageFileProcessor.SheetContent;
import one.edee.babylon.export.stats.MessageFileExportStats;
import one.edee.babylon.snapshot.TranslationSnapshotReadContract;
import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;

public class MessageFileProcessorTest {

    @Test
    public void testWhenAllTranslationsExistAndUnchanged_thenNotIncludedInSheet() {
        Map<String, String> keys1 = new HashMap<>();
        keys1.put("pagination.prev", "Previous");
        keys1.put("pagination.next", "Next");

        Map<String, Map<String, String>> map1 = new HashMap<>();
        map1.put("i18n/common.properties", keys1);
        TranslationSnapshotReadContract snapshot = new FakeTranslationSnapshot(map1);

        Map<String, String> primaryMessages = new HashMap<>();
        keys1.put("pagination.prev", "Previous");
        keys1.put("pagination.next", "Next");

        List<String> translateTo = Arrays.asList("cz", "sk");

        Map<String, String> keysCz = new HashMap<>();
        keys1.put("pagination.prev", "Předchozí");
        keys1.put("pagination.next", "Následující");
        Map<String, String> keysSk = new HashMap<>();
        keys1.put("pagination.prev", "Predchádzajúce");
        keys1.put("pagination.next", "Nasledujúce");

        Map<String, Map<String, String>> translations = new HashMap<>();
        translations.put("cz", keysCz);
        translations.put("sk", keysSk);

        SheetContent expected = new SheetContent(
                Arrays.asList("key", "primary", "cz", "sk"),
                Collections.emptyList()
        );

        MessageFileProcessor mfProcessor = new MessageFileProcessor(snapshot);
        SheetContent sheet = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo).getFirst();

        assertEquals(expected, sheet);
    }

    @Test
    public void whenNoSnapshotOfPreviousTranslationAndNoTranslationsThenAllMessagesAreExportedToTranslationSheet() {
        FakeTranslationSnapshot emptySnapshot = new FakeTranslationSnapshot(new HashMap<>());

        Map<String, String> primaryMessages = BMap.of(
                "pagination.prev", "Previous",
                "pagination.next", "Next",
                "price.zeroPriceString", "Free");

        List<String> translateTo = Arrays.asList("cz", "sk");
        Map<String, Map<String, String>> noTranslations = new HashMap<>();

        SheetContent expected = new SheetContent(
                Arrays.asList("key", "primary", "cz", "sk"),
                Arrays.asList(
                        Arrays.asList("pagination.prev", "Previous", null, null),
                        Arrays.asList("pagination.next", "Next", null, null),
                        Arrays.asList("price.zeroPriceString", "Free", null, null)
                )
        );

        MessageFileProcessor mfProcessor = new MessageFileProcessor(emptySnapshot);
        Pair<SheetContent, MessageFileExportStats> result = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, noTranslations, translateTo);
        assertEquals(expected, result.getFirst());
    }

    @Test
    public void whenThereIsANewPrimaryMessageFileThenItsMessagesAreExportedToTranslationSheetButTheMessagesAlreadyInSnapshotAreNot() {
        FakeTranslationSnapshot snapshot = new FakeTranslationSnapshot(BMap.of(
                "i18n/common.properties", BMap.of(
                        "pagination.prev", "Previous",
                        "pagination.next", "Next",
                        "price.zeroPriceString", "Free")));

        Map<String, String> primaryMessages = BMap.of(
                "pagination.prev", "Previous",
                "pagination.next", "Next",
                "price.zeroPriceString", "Free",
                "availability.text.ALWAYS_AVAILABLE", "Always in stock",
                "availability.text.NOT_AVAILABLE", "Momentarily unavailable");

        List<String> translateTo = Arrays.asList("cz", "sk");
        Map<String, Map<String, String>> translations = BMap.of(
                "cz", BMap.of(
                        "pagination.prev", "Předchozí",
                        "pagination.next", "Následující",
                        "price.zeroPriceString", "Zdarma"),
                "sk", BMap.of(
                        "pagination.prev", "Predchádzajúce",
                        "pagination.next", "Nasledujúce",
                        "price.zeroPriceString", "Zadarmo"));

        SheetContent expected = new SheetContent(
                Arrays.asList("key", "primary", "cz", "sk"),
                Arrays.asList(
                        Arrays.asList("availability.text.ALWAYS_AVAILABLE", "Always in stock", null, null),
                        Arrays.asList("availability.text.NOT_AVAILABLE", "Momentarily unavailable", null, null)
                )
        );

        MessageFileProcessor mfProcessor = new MessageFileProcessor(snapshot);
        Pair<SheetContent, MessageFileExportStats> result = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo);
        assertEquals(expected, result.getFirst());
    }

    @Test
    public void whenMessageIsMissingSomeTranslationsThenTheMessageValuesExistingTranslationsAreIncludedInTheTranslationSheet() {
        FakeTranslationSnapshot emptySnapshot = new FakeTranslationSnapshot(new HashMap<>());

        Map<String, String> primaryMessages = BMap.of(
                "pagination.prev", "Previous",
                "pagination.next", "Next",
                "price.zeroPriceString", "Free");

        List<String> translateTo = Arrays.asList("cz", "sk");
        Map<String, Map<String, String>> translations = BMap.of(
                "cz", BMap.of(
                        "pagination.prev", "Předchozí"),
                "sk", BMap.of(
                        "pagination.next", "Nasledujúce")
        );

        SheetContent expected = new SheetContent(
                Arrays.asList("key", "primary", "cz", "sk"),
                Arrays.asList(
                        Arrays.asList("pagination.prev", "Previous", "Předchozí", null),
                        Arrays.asList("pagination.next", "Next", null, "Nasledujúce"),
                        Arrays.asList("price.zeroPriceString", "Free", null, null)
                )
        );

        MessageFileProcessor mfProcessor = new MessageFileProcessor(emptySnapshot);
        Pair<SheetContent, MessageFileExportStats> result = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo);
        assertEquals(expected, result.getFirst());
    }

    @Test
    public void whenMessageIsPresentInSnapshotButIsMissingFromPrimaryMessageFileThenItIsNotExportedToTranslationSheet() {
        FakeTranslationSnapshot snapshot = new FakeTranslationSnapshot(BMap.of(
                "i18n/common.properties", BMap.of(
                        "pagination.prev", "Previous",
                        "pagination.next", "Next",
                        "price.zeroPriceString", "Free")));

        Map<String, String> primaryMessages = BMap.of(
                "pagination.prev", "Previous",
                "pagination.next", "Next",
                "availability.text.ALWAYS_AVAILABLE", "Always in stock",
                "availability.text.NOT_AVAILABLE", "Momentarily unavailable");

        List<String> translateTo = Arrays.asList("cz", "sk");
        Map<String, Map<String, String>> translations = BMap.of(
                "cz", BMap.of(
                        "pagination.prev", "Předchozí",
                        "pagination.next", "Následující"),
                "sk", BMap.of(
                        "pagination.prev", "Predchádzajúce",
                        "pagination.next", "Nasledujúce"));

        SheetContent expected = new SheetContent(
                Arrays.asList("key", "primary", "cz", "sk"),
                Arrays.asList(
                        Arrays.asList("availability.text.ALWAYS_AVAILABLE", "Always in stock", null, null),
                        Arrays.asList("availability.text.NOT_AVAILABLE", "Momentarily unavailable", null, null)
                )
        );

        MessageFileProcessor mfProcessor = new MessageFileProcessor(snapshot);
        Pair<SheetContent, MessageFileExportStats> result = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo);
        assertEquals(expected, result.getFirst());
    }

    @Test
    public void whenMessageDiffersFromTheValueInSnapshotThenAllTranslationsForThisMessageAreEmptyInTheTranslationSheet() {
        FakeTranslationSnapshot snapshot = new FakeTranslationSnapshot(BMap.of(
                "i18n/common.properties", BMap.of(
                        "pagination.prev", "Previous",
                        "pagination.next", "Next",
                        "price.zeroPriceString", "Free")));

        Map<String, String> primaryMessages = BMap.of(
                "pagination.prev", "Prev",
                "pagination.next", "Next",
                "price.zeroPriceString", "Free");

        List<String> translateTo = Arrays.asList("cz", "sk");
        Map<String, Map<String, String>> translations = BMap.of(
                "cz", BMap.of(
                        "pagination.prev", "Předchozí",
                        "pagination.next", "Následujúce"),
                "sk", BMap.of(
                        "pagination.prev", "Predchádzajúce",
                        "pagination.next", "Nasledujúce"));

        SheetContent expected = new SheetContent(
                Arrays.asList("key", "primary", "cz", "sk"),
                Arrays.asList(
                        Arrays.asList("pagination.prev", "Prev", null, null),
                        Arrays.asList("price.zeroPriceString", "Free", null, null)
                )
        );

        MessageFileProcessor mfProcessor = new MessageFileProcessor(snapshot);
        Pair<SheetContent, MessageFileExportStats> result = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo);
        assertEquals(expected, result.getFirst());
    }

    @Test
    public void whenTranslationIsMissingAValueForALanguageThenTheMissingValueIsInCorrectPlace() {
        FakeTranslationSnapshot snapshot = new FakeTranslationSnapshot(BMap.of(
                "i18n/errors.properties", BMap.of(
                        "error.command.securityBreach", "Zdá se, že jste delší dobu nebyli aktivní. Zkontrolujte, zda jste nebyli automaticky odhlášeni a opět se přihlaste nebo případně vyplňte formulář znovu.",
                        "error.exceptionhandler.missingMandatoryValue", "Vyžadována povinná hodnota, ale ta chybí!")));

        Map<String, String> primaryMessages = BMap.of(
                "error.command.securityBreach", "Zdá se, že jste delší dobu nebyli aktivní. Zkontrolujte, zda jste nebyli automaticky odhlášeni a opět se přihlaste nebo případně vyplňte formulář znovu.",
                "error.exceptionhandler.missingMandatoryValue", "Vyžadována povinná hodnota, ale ta chybí!");

        List<String> translateTo = Arrays.asList("de", "sk");
        Map<String, Map<String, String>> translations = BMap.of(
                "de", BMap.of(
                        "error.exceptionhandler.missingMandatoryValue", "Pflichtfeld erforderlich, fehlt aber!"),
                "sk", BMap.of(
                        "error.command.securityBreach", "Zdá sa, že ste dlhšiu dobu neboli aktívni. Skontrolujte, či ste neboli automaticky odhlásení, a opäť sa prihláste, prípadne vyplňte formulár odznova.",
                        "error.exceptionhandler.missingMandatoryValue", "Vyžaduje sa povinná hodnota, ale tá chýba!"));

        SheetContent expected = new SheetContent(
                Arrays.asList("key", "primary", "de", "sk"),
                Collections.singletonList(
                        Arrays.asList("error.command.securityBreach",
                                "Zdá se, že jste delší dobu nebyli aktivní. Zkontrolujte, zda jste nebyli automaticky odhlášeni a opět se přihlaste nebo případně vyplňte formulář znovu.",
                                null,
                                "Zdá sa, že ste dlhšiu dobu neboli aktívni. Skontrolujte, či ste neboli automaticky odhlásení, a opäť sa prihláste, prípadne vyplňte formulár odznova.")
                )
        );

        MessageFileProcessor mfProcessor = new MessageFileProcessor(snapshot);
        Pair<SheetContent, MessageFileExportStats> result = mfProcessor.prepareTranslationSheet("i18n/errors.properties", primaryMessages, translations, translateTo);
        assertEquals(expected, result.getFirst());
    }

    private static class BMap {
        public static <T> Map<String, T> of(Object... values) {
            Map<String, T> map = new HashMap<>();
            for (int i = 0; i < values.length; i++) {
                if (i % 2 == 0) {
                    String key = (String) values[i];
                    T value = (T) values[i + 1];
                    map.put(key, value);
                }
            }


            return map;
        }

    }
}

File: src/test/kotlin/com/fg/util/babylon/export/MessageFileProcessorTest.kt
package com.fg.util.babylon.export

import junit.framework.Assert.assertEquals
import org.junit.Test

class MessageFileProcessorTest {

    @Test
    fun `when a message has translations for all desired languages and has not changed from snaphost then it is not included in the translation sheet`() {
        val snapshot = FakeTranslationSnapshot(mapOf(
                "i18n/common.properties" to mapOf(
                        "pagination.prev" to "Previous",
                        "pagination.next" to "Next")))

        val primaryMessages = mapOf(
                "pagination.prev" to "Previous",
                "pagination.next" to "Next")

        val translateTo = listOf("cz", "sk")
        val translations = mapOf(
                "cz" to mapOf(
                        "pagination.prev" to "Předchozí",
                        "pagination.next" to "Následující"),
                "sk" to mapOf(
                        "pagination.prev" to "Predchádzajúce",
                        "pagination.next" to "Nasledujúce"))

        val expected = SheetContent(
                listOf("key", "primary", "cz", "sk"),
                emptyList()
        )

        val mfProcessor = MessageFileProcessor(snapshot)
        val (sheet, _) = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo)
        assertEquals(expected, sheet)
    }

    @Test
    fun `when no snapshot of previous translation and no translations then all messages are exported to translation sheet`() {
        val emptySnapshot = FakeTranslationSnapshot(emptyMap())

        val primaryMessages = mapOf(
                "pagination.prev" to "Previous",
                "pagination.next" to "Next",
                "price.zeroPriceString" to "Free")

        val translateTo = listOf("cz", "sk")
        val noTranslations = emptyMap<Language, Messages>()

        val expected = SheetContent(
                listOf("key", "primary", "cz", "sk"),
                listOf(
                        listOf("pagination.prev", "Previous", null, null),
                        listOf("pagination.next", "Next", null, null),
                        listOf("price.zeroPriceString", "Free", null, null)
                )
        )

        val mfProcessor = MessageFileProcessor(emptySnapshot)
        val (sheet, _) = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, noTranslations, translateTo)
        assertEquals(expected, sheet)
    }

    @Test
    fun `when there is a new primary message file then its messages are exported to translation sheet, but the messages already in snapshot are not`() {
        val snapshot = FakeTranslationSnapshot(
                mapOf("i18n/common.properties" to mapOf(
                        "pagination.prev" to "Previous",
                        "pagination.next" to "Next",
                        "price.zeroPriceString" to "Free")))

        val primaryMessages = mapOf(
                "pagination.prev" to "Previous",
                "pagination.next" to "Next",
                "price.zeroPriceString" to "Free",
                "availability.text.ALWAYS_AVAILABLE" to "Always in stock",
                "availability.text.NOT_AVAILABLE" to "Momentarily unavailable")

        val translateTo = listOf("cz", "sk")
        val translations = mapOf(
                "cz" to mapOf(
                        "pagination.prev" to "Předchozí",
                        "pagination.next" to "Následující",
                        "price.zeroPriceString" to "Zdarma"),
                "sk" to mapOf(
                        "pagination.prev" to "Predchádzajúce",
                        "pagination.next" to "Nasledujúce",
                        "price.zeroPriceString" to "Zadarmo"))

        val expected = SheetContent(
                listOf("key", "primary", "cz", "sk"),
                listOf(
                        listOf("availability.text.ALWAYS_AVAILABLE", "Always in stock", null, null),
                        listOf("availability.text.NOT_AVAILABLE", "Momentarily unavailable", null, null)
                )
        )

        val mfProcessor = MessageFileProcessor(snapshot)
        val (sheet, _) = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo)
        assertEquals(expected, sheet)
    }

    @Test
    fun `when a message is missing some translations then the message values existing translations are included in the translation sheet`() {
        val emptySnapshot = FakeTranslationSnapshot(emptyMap())

        val primaryMessages = mapOf(
                "pagination.prev" to "Previous",
                "pagination.next" to "Next",
                "price.zeroPriceString" to "Free")

        val translateTo = listOf("cz", "sk")
        val translations = mapOf(
                "cz" to mapOf(
                        "pagination.prev" to "Předchozí"),
                "sk" to mapOf(
                        "pagination.next" to "Nasledujúce")
        )

        val expected = SheetContent(
                listOf("key", "primary", "cz", "sk"),
                listOf(
                        listOf("pagination.prev", "Previous", "Předchozí", null),
                        listOf("pagination.next", "Next", null, "Nasledujúce"),
                        listOf("price.zeroPriceString", "Free", null, null)
                )
        )

        val mfProcessor = MessageFileProcessor(emptySnapshot)
        val (sheet, _) = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo)
        assertEquals(expected, sheet)
    }

    @Test
    fun `when a message is present in snapshot but is missing from primary message file then it is not exported to translation sheet`() {
        val snapshot = FakeTranslationSnapshot(mapOf(
                "i18n/common.properties" to mapOf(
                        "pagination.prev" to "Previous",
                        "pagination.next" to "Next",
                        "price.zeroPriceString" to "Free")))

        val primaryMessages = mapOf(
                "pagination.prev" to "Previous",
                "pagination.next" to "Next",
                "availability.text.ALWAYS_AVAILABLE" to "Always in stock",
                "availability.text.NOT_AVAILABLE" to "Momentarily unavailable")

        val translateTo = listOf("cz", "sk")
        val translations = mapOf(
                "cz" to mapOf(
                        "pagination.prev" to "Předchozí",
                        "pagination.next" to "Následující"),
                "sk" to mapOf(
                        "pagination.prev" to "Predchádzajúce",
                        "pagination.next" to "Nasledujúce"))

        val expected = SheetContent(
                listOf("key", "primary", "cz", "sk"),
                listOf(
                        listOf("availability.text.ALWAYS_AVAILABLE", "Always in stock", null, null),
                        listOf("availability.text.NOT_AVAILABLE", "Momentarily unavailable", null, null)
                )
        )

        val mfProcessor = MessageFileProcessor(snapshot)
        val (sheet, _) = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo)
        assertEquals(expected, sheet)
    }

    @Test
    fun `when a message differs from the value in snapshot then all translations for this messages are empty in the translation sheet`() {
        val snapshot = FakeTranslationSnapshot(mapOf(
                "i18n/common.properties" to mapOf(
                        "pagination.prev" to "Previous",
                        "pagination.next" to "Next",
                        "price.zeroPriceString" to "Free")))

        val primaryMessages = mapOf(
                "pagination.prev" to "Prev",
                "pagination.next" to "Next",
                "price.zeroPriceString" to "Free")

        val translateTo = listOf("cz", "sk")
        val translations = mapOf(
                "cz" to mapOf(
                        "pagination.prev" to "Předchozí",
                        "pagination.next" to "Následující"),
                "sk" to mapOf(
                        "pagination.prev" to "Predchádzajúce",
                        "pagination.next" to "Nasledujúce"))

        val expected = SheetContent(
                listOf("key", "primary", "cz", "sk"),
                listOf(
                        listOf("pagination.prev", "Prev", null, null),
                        listOf("price.zeroPriceString", "Free", null, null)
                )
        )

        val mfProcessor = MessageFileProcessor(snapshot)
        val (sheet, _) = mfProcessor.prepareTranslationSheet("i18n/common.properties", primaryMessages, translations, translateTo)
        assertEquals(expected, sheet)
    }

    @Test
    fun `when translation is missing a value for a language then the missing value is in correct place`() {
        val snapshot = FakeTranslationSnapshot(mapOf(
                "i18n/errors.properties" to mapOf(
                        "error.command.securityBreach" to "Zdá se, že jste delší dobu nebyli aktivní. Zkontrolujte, zda jste nebyli automaticky odhlášeni a opět se přihlaste nebo případně vyplňte formulář znovu.",
                        "error.exceptionhandler.missingMandatoryValue" to "Vyžadována povinná hodnota, ale ta chybí!")))

        val primaryMessages = mapOf(
                "error.command.securityBreach" to "Zdá se, že jste delší dobu nebyli aktivní. Zkontrolujte, zda jste nebyli automaticky odhlášeni a opět se přihlaste nebo případně vyplňte formulář znovu.",
                "error.exceptionhandler.missingMandatoryValue" to "Vyžadována povinná hodnota, ale ta chybí!")

        val translateTo = listOf("de", "sk")
        val translations = mapOf(
                "de" to mapOf(
                        "error.exceptionhandler.missingMandatoryValue" to "Pflichtfeld erforderlich, fehlt aber!"),
                "sk" to mapOf(
                        "error.command.securityBreach" to "Zdá sa, že ste dlhšiu dobu neboli aktívni. Skontrolujte, či ste neboli automaticky odhlásení, a opäť sa prihláste, prípadne vyplňte formulár odznova.",
                        "error.exceptionhandler.missingMandatoryValue" to "Vyžaduje sa povinná hodnota, ale tá chýba!"))

        val expected = SheetContent(
                listOf("key", "primary", "de", "sk"),
                listOf(
                listOf("error.command.securityBreach",
                        "Zdá se, že jste delší dobu nebyli aktivní. Zkontrolujte, zda jste nebyli automaticky odhlášeni a opět se přihlaste nebo případně vyplňte formulář znovu.",
                        null,
                        "Zdá sa, že ste dlhšiu dobu neboli aktívni. Skontrolujte, či ste neboli automaticky odhlásení, a opäť sa prihláste, prípadne vyplňte formulár odznova.")
                )
        )

        val mfProcessor = MessageFileProcessor(snapshot)
        val (sheet, _) = mfProcessor.prepareTranslationSheet("i18n/errors.properties", primaryMessages, translations, translateTo)
        assertEquals(expected, sheet)
    }

}